pkg/controller/machine/controller.go

```diff
@@ -1242,3 +1259,46 @@ func (r *Reconciler) updateNode(ctx context.Context, node *corev1.Node, modifier
 		return r.client.Update(ctx, node)
 	})
 }
+
+// handleNodeFailuresWithExternalCCM reacts to node status discovery of CCM's node lifecycle controller.
+// If an instance at cloud provider is not found then it waits till CCM deletes node objects, that allows:
+//   - create a new instance at cloud provider
+//   - initialize a new node object - the object should not be reused between instance creation
+//     for example, instance foo that got deleted and recreated should initialize a completely new node object
+//     instead of reusing the old one as it can cause problems to update node's metadata, like IP address.
+//
+// If node is shut-down it allows MC to react accordingly to specific cloud provider requirements, those are:
+//   - wait for node to become online again or
+//   - delete a machine which cannot be recovered
+func (r *Reconciler) handleNodeFailuresWithExternalCCM(
+	ctx context.Context,
+	log *zap.SugaredLogger,
+	prov cloudprovidertypes.Provider,
+	provConfig *providerconfigtypes.Config,
+	node *corev1.Node,
+	machine *clusterv1alpha1.Machine,
+) (*reconcile.Result, error) {
+	taintShutdown := corev1.Taint{
+		Key:    ccmapi.TaintNodeShutdown,
+		Effect: corev1.TaintEffectNoSchedule,
+	}
+
+	_, err := prov.Get(ctx, log, machine, r.providerData)
+	if err != nil {
+		if cloudprovidererrors.IsNotFound(err) {
+			log.Info("The node does not have corresponding instance, waiting for CCM to delete it")
+			return &reconcile.Result{RequeueAfter: deletionRetryWaitPeriod}, nil
+		}
+		return nil, err
+	} else if taintExists(node, taintShutdown) {
+		switch provConfig.CloudProvider {
```

Comment:
We should check other supported cloud providers regarding instance shutdown, if it can be powered on safely later or we should delete machine and recreate.

Reply:
>  if it can be powered on safely later or we should delete machine and recreate.

Was this meant to say "if it cannot be powered on safely later"?

Reply:
right

Reply:
> Was this meant to say "if it cannot be powered on safely later"?

Nope, I meant if machine is shutdown and powered on later.

Reply:
AFAIK some CCMs when discover that instance has been shutdown remove nodes object others don't. I remember OpenStack CCM in the past was deleting node object once you powered off the machine. We have to check if this is still a case for providers we support. However, that should be a different PR and issue.